src/Week2/LogAnalysis.hs

```diff
@@ -0,0 +1,133 @@
+  | t2 > t1   = Node (insert msg1 left) msg2 right
+  | t2 < t1   = Node left msg2 (insert msg1 right)
+  | otherwise = tree
+
+
+{-
+ Exercise 3: Once we can insert a single LogMessage into a MessageTree,
+ we can build a complete MessageTree from a list of messages. Specifically, define a function
+    build :: [LogMessage] -> MessageTree
+
+which builds up a MessageTree containing the messages in the list,
+ by successively
+-}
+build :: [LogMessage] -> MessageTree
+build = foldr insert Leaf
+
+
+{-
+ Exercise 4: Finally, define the function
+   inOrder :: MessageTree -> [LogMessage]
+
+ which takes a sorted MessageTree and produces a list of all the
+ LogMessages it contains, sorted by timestamp from smallest to biggest.
+ (This is known as an in-order traversal of the MessageTree.)
+
+ With these functions, we can now remove Unknown messages and
+ sort the well-formed messages using an expression such as:
+   inOrder (build tree)
+-}
+inOrder :: MessageTree -> [LogMessage]
+-- base case
+inOrder Leaf                  = []
+-- Visit in the order: Left subtree -> root -> right subtree
+inOrder (Node left msg right) = inOrder left ++ [msg] ++ inOrder right
+
+
+{-
+ Exercise 5: Now that we can sort the log messages, the only thing
+ left to do is extract the relevant information. We have decided that
+ “relevant” means “errors with a severity of at least 50”.
+
+ Write a function
+    whatWentWrong :: [LogMessage] -> [String]
+ which takes an unsorted list of LogMessages, and returns a list of the
+ messages corresponding to any errors with a severity of 50 or greater,
+ sorted by timestamp. (Of course, you can use your functions from the
+ previous exercises to do the sorting.)
+
+ For example, suppose our log file looked like this:
+  I 6 Completed armadillo processing
+  I 1 Nothing to report
+  E 99 10 Flange failed!
+  I 4 Everything normal
+  I 11 Initiating self-destruct sequence
+  E 70 3 Way too many pickles
+  E 65 8 Bad pickle-flange interaction detected
+  W 5 Flange is due for a check-up
+  I 7 Out for lunch, back in two time steps
+  E 20 2 Too many pickles
+  I 9 Back from lunch
+
+ This file is provided as sample.log. There are four errors, three of
+ which have a severity of greater than 50. The output of whatWentWrong
+ on sample.log ought to be
+
+  [ "Way too many pickles"
+    , "Bad pickle-flange interaction detected"
+    , "Flange failed!"
+  ]
+
+ You can test your whatWentWrong function with testWhatWentWrong,
+ which is also provided by the Log module. You should provide
+ testWhatWentWrong with your parse function, your whatWentWrong
+ function, and the name of the log file to parse.
+-}
+whatWentWrong :: [LogMessage] -> [String]
+whatWentWrong msgs = map getMsg $ filter important $ (inOrder . build) msgs
+    where important (LogMessage (Error s) _ _) = s >= 50
+          important _ = False
+          getMsg (LogMessage _ _ msg) = msg
+          getMsg _                    = ""
```

Comment:
doing `map getMsg` will lead to final list containing many `""`-s.
Can you think of another way to implement this so that empty lists are not included?

<details>
  <summary>hint</summary>
  <pre>
https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Maybe.html#v:mapMaybe
  </pre>
</details>

Reply:
So, what I understood is that we build a list of `Just` and `Nothing` and using `mapMaybe` filter out the `Just` values.
Can you please have a look and see if that's what you meant?

Reply:
Yes, that's exactly what I had in mind. mapMaybe allows you to express mapping and filtering using just one function.